File: dia-cli/src/bookmarks.rs
use anyhow::{Context, Result};
use serde::Deserialize;
use std::fs::File;
use std::io::BufReader;
use std::path::Path;

use crate::model::Entry;

#[derive(Deserialize)]
struct BookmarkFile {
    roots: BookmarkRoots,
}

#[derive(Deserialize)]
struct BookmarkRoots {
    bookmark_bar: Option<BookmarkNode>,
    other: Option<BookmarkNode>,
    synced: Option<BookmarkNode>,
}

#[derive(Deserialize)]
struct BookmarkNode {
    name: Option<String>,
    #[serde(rename = "type")]
    node_type: Option<String>,
    url: Option<String>,
    children: Option<Vec<BookmarkNode>>,
}

const MAX_BOOKMARKS: usize = 10000;

pub fn load_bookmarks(bookmarks_path: &Path) -> Result<Vec<Entry>> {
    if !bookmarks_path.exists() {
        return Ok(Vec::new());
    }

    let file = File::open(bookmarks_path)
        .with_context(|| format!("failed to open bookmarks at {}", bookmarks_path.display()))?;

    let reader = BufReader::with_capacity(16 * 1024, file);
    let bookmark_file: BookmarkFile = serde_json::from_reader(reader)
        .with_context(|| format!("failed to parse bookmarks JSON at {}", bookmarks_path.display()))?;

    let mut entries = Vec::with_capacity(500);

    if let Some(ref node) = bookmark_file.roots.bookmark_bar {
        flatten_node(node, String::new(), &mut entries);
    }
    if let Some(ref node) = bookmark_file.roots.other {
        flatten_node(node, String::new(), &mut entries);
    }
    if let Some(ref node) = bookmark_file.roots.synced {
        flatten_node(node, String::new(), &mut entries);
    }

    Ok(entries)
}

fn flatten_node(node: &BookmarkNode, folder_path: String, entries: &mut Vec<Entry>) {
    if entries.len() >= MAX_BOOKMARKS {
        return;
    }

    let node_type = node.node_type.as_deref().unwrap_or("unknown");

    match node_type {
        "url" => {
            if let (Some(url), Some(title)) = (&node.url, &node.name) {
                let folder = if folder_path.is_empty() {
                    None
                } else {
                    Some(folder_path)
                };
                entries.push(Entry::new_bookmark(url.clone(), title.clone(), folder));
            }
        }
        "folder" => {
            let new_path = match (&folder_path, &node.name) {
                (p, Some(n)) if p.is_empty() => n.clone(),
                (p, Some(n)) => format!("{} / {}", p, n),
                (p, None) => p.clone(),
            };

            if let Some(children) = &node.children {
                for child in children {
                    flatten_node(child, new_path.clone(), entries);
                }
            }
        }
        _ => {}
    }
}

File: dia-cli/src/bookmarks.zig
const std = @import("std");
const model = @import("model.zig");

const Entry = model.Entry;

const BookmarkFile = struct {
    roots: BookmarkRoots,
};

const BookmarkRoots = struct {
    bookmark_bar: ?BookmarkNode = null,
    other: ?BookmarkNode = null,
    synced: ?BookmarkNode = null,
};

const BookmarkNode = struct {
    name: ?[]const u8 = null,
    type: ?[]const u8 = null,
    url: ?[]const u8 = null,
    children: ?[]BookmarkNode = null,
};

const MAX_BOOKMARKS = 10_000;

pub fn loadBookmarks(allocator: std.mem.Allocator, path: []const u8) ![]Entry {
    var file = std.fs.openFileAbsolute(path, .{}) catch |err| switch (err) {
        error.FileNotFound => return try allocator.alloc(Entry, 0),
        else => return err,
    };
    defer file.close();

    const data = try file.readToEndAlloc(allocator, 16 * 1024 * 1024);
    defer allocator.free(data);

    var parsed = try std.json.parseFromSlice(BookmarkFile, allocator, data, .{
        .ignore_unknown_fields = true,
    });
    defer parsed.deinit();

    var entries = std.ArrayListUnmanaged(Entry){};
    errdefer entries.deinit(allocator);

    if (parsed.value.roots.bookmark_bar) |node| {
        try flattenNode(allocator, node, "", &entries);
    }
    if (parsed.value.roots.other) |node| {
        try flattenNode(allocator, node, "", &entries);
    }
    if (parsed.value.roots.synced) |node| {
        try flattenNode(allocator, node, "", &entries);
    }

    return entries.toOwnedSlice(allocator);
}

fn flattenNode(
    allocator: std.mem.Allocator,
    node: BookmarkNode,
    folder_path: []const u8,
    entries: *std.ArrayListUnmanaged(Entry),
) !void {
    if (entries.items.len >= MAX_BOOKMARKS) return;

    const node_type = node.type orelse "unknown";
    if (std.mem.eql(u8, node_type, "url")) {
        if (node.url) |url| {
            if (node.name) |title| {
                const folder = if (folder_path.len == 0) null else folder_path;
                try entries.append(allocator, try Entry.initBookmark(allocator, url, title, folder));
            }
        }
        return;
    }

    if (!std.mem.eql(u8, node_type, "folder")) return;

    const path_for_children = try buildFolderPath(allocator, folder_path, node.name);
    defer allocator.free(path_for_children);

    if (node.children) |children| {
        for (children) |child| {
            try flattenNode(allocator, child, path_for_children, entries);
        }
    }
}

fn buildFolderPath(
    allocator: std.mem.Allocator,
    base: []const u8,
    name: ?[]const u8,
) ![]u8 {
    if (name) |n| {
        if (base.len == 0) {
            return allocator.dupe(u8, n);
        }
        return std.fmt.allocPrint(allocator, "{s} / {s}", .{ base, n });
    }
    return allocator.dupe(u8, base);
}

// tests
fn writeFixture(dir: std.fs.Dir, name: []const u8, content: []const u8) !void {
    try dir.writeFile(.{ .sub_path = name, .data = content });
}

test "load bookmarks basic" {
    var tmp = std.testing.tmpDir(.{});
    defer tmp.cleanup();
    const dir_path = try tmp.dir.realpathAlloc(std.testing.allocator, ".");
    defer std.testing.allocator.free(dir_path);
    const path = try std.fs.path.join(std.testing.allocator, &.{ dir_path, "Bookmarks" });
    defer std.testing.allocator.free(path);

    const json =
        \\{
        \\  "roots": {
        \\    "bookmark_bar": {
        \\      "type": "folder",
        \\      "name": "Bookmarks Bar",
        \\      "children": [
        \\        {"type": "url", "url": "https://example.com", "name": "Example"}
        \\      ]
        \\    },
        \\    "other": {"type": "folder", "children": []},
        \\    "synced": {"type": "folder", "children": []}
        \\  }
        \\}
    ;
    try writeFixture(tmp.dir, "Bookmarks", json);

    var arena = std.heap.ArenaAllocator.init(std.testing.allocator);
    defer arena.deinit();
    const alloc = arena.allocator();
    const entries = try loadBookmarks(alloc, path);
    try std.testing.expectEqual(@as(usize, 1), entries.len);
    try std.testing.expectEqualStrings("https://example.com", entries[0].url);
    try std.testing.expectEqualStrings("Example", entries[0].title);
}

test "load bookmarks nested folders" {
    var tmp = std.testing.tmpDir(.{});
    defer tmp.cleanup();
    const dir_path = try tmp.dir.realpathAlloc(std.testing.allocator, ".");
    defer std.testing.allocator.free(dir_path);
    const path = try std.fs.path.join(std.testing.allocator, &.{ dir_path, "Bookmarks" });
    defer std.testing.allocator.free(path);

    const json =
        \\{
        \\  "roots": {
        \\    "bookmark_bar": {
        \\      "type": "folder",
        \\      "name": "Bar",
        \\      "children": [
        \\        {
        \\          "type": "folder",
        \\          "name": "Work",
        \\          "children": [
        \\            {"type": "url", "url": "https://jira.com", "name": "Jira"}
        \\          ]
        \\        }
        \\      ]
        \\    }
        \\  }
        \\}
    ;
    try writeFixture(tmp.dir, "Bookmarks", json);

    var arena = std.heap.ArenaAllocator.init(std.testing.allocator);
    defer arena.deinit();
    const alloc = arena.allocator();
    const entries = try loadBookmarks(alloc, path);
    try std.testing.expectEqualStrings("Bar / Work", entries[0].folder.?);
}

test "load bookmarks missing file returns empty" {
    var arena = std.heap.ArenaAllocator.init(std.testing.allocator);
    defer arena.deinit();
    const alloc = arena.allocator();
    const entries = try loadBookmarks(alloc, "/nonexistent/bookmarks");
    try std.testing.expectEqual(@as(usize, 0), entries.len);
}
